### PR TITLE
Add MaximumMeanDiscrepancy metric

### PR DESCRIPTION
**Pull request by kzkadc**
Description: added maximum mean discrepancy (MMD) metric, which measures the discrepancy of two distributions.
https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR, @kzkadc 
Right now CI is failing with related broken tests, can you please check them ?

File: ignite/metrics/maximum_mean_discrepancy.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+class MaximumMeanDiscrepancy(Metric):
+    r"""Calculates the mean of `maximum mean discrepancy (MMD)
+    <https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html>`_.
+
+    .. math::
+       \begin{align*}
+           \text{MMD}^2 (P,Q) &= \underset{\| f \| \leq 1}{\text{sup}} | \mathbb{E}_{X\sim P}[f(X)]
+           - \mathbb{E}_{Y\sim Q}[f(Y)] |^2 \\
+           &\approx \frac{1}{B(B-1)} \sum_{i} \sum_{j\neq i}k(\mathbf{x}_i,\mathbf{x}_j)
```
Comment (vfdev-5) [inline ignite/metrics/maximum_mean_discrepancy.py:19]: Is it possible to specify i, j ranges?

File: ignite/metrics/maximum_mean_discrepancy.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        n = x.shape[0]
+        XX = (XX.sum() - n) / (n * (n - 1))
+        YY = (YY.sum() - n) / (n * (n - 1))
+        XY = XY.sum() / (n * n)
+
+        # mmd cannot be negative
+        mmd2 = (XX - 2.0 * XY + YY).clamp(min=0.0)
+
+        self._sum_of_mmd += mmd2.sqrt().to(self._device)
+        self._num_batches += 1
```
Comment (vfdev-5) [inline ignite/metrics/maximum_mean_discrepancy.py:125]: will resulting metric value practically depend on the batch size?

Comment (kzkadc) [reply ignite/metrics/maximum_mean_discrepancy.py:125]: Thanks for your review.
Yes, it depends on the batch size.
But mmd2 above is an unbiased estimator of MMD^2 ( https://jmlr.org/papers/volume13/gretton12a/gretton12a.pdf , Lemma 6 ).

I will modify the code to accumulate mmd2 and take the sqrt in compute().
Then, we can expect that the computed mmd will converge with a sufficient number of batches.

File: ignite/metrics/maximum_mean_discrepancy.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    .. math::
+       \begin{align*}
+           \text{MMD}^2 (P,Q) &= \underset{\| f \| \leq 1}{\text{sup}} | \mathbb{E}_{X\sim P}[f(X)]
+           - \mathbb{E}_{Y\sim Q}[f(Y)] |^2 \\
+           &\approx \frac{1}{B(B-1)} \sum_{i=1}^B \sum_{\substack{j=1 \\ j\neq i}}^B k(\mathbf{x}_i,\mathbf{x}_j)
+           -\frac{2}{B^2}\sum_{i=1}^B \sum_{j=1}^B k(\mathbf{x}_i,\mathbf{y}_j)
+           + \frac{1}{B(B-1)} \sum_{i=1}^B \sum_{\substack{j=1 \\ j\neq i}}^B k(\mathbf{y}_i,\mathbf{y}_j)
+       \end{align*}
+
+    where :math:`B` is the batch size, and :math:`\mathbf{x}_i` and :math:`\mathbf{y}_j` are
```
Comment (vfdev-5) [inline ignite/metrics/maximum_mean_discrepancy.py:24]: Sorry to coming back to this question, actually, I'm not quite sure about that B is the batch size.
According to https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html, their `m` is the num of samples.
Also, https://jmlr.org/papers/volume13/gretton12a/gretton12a.pdf sets the problem as 
> Let x and y be random variables defined on a topological space X, with respective
Borel probability measures p and q . Given observations X := {x1, . . . , xm} and Y := {y1, . . . , yn},
independently and identically distributed (i.i.d.) from p and q, respectively, can we decide whether
p != q?

where m and n are num of samples of the observations.

Should not we use the same idea and replace B per N = num of samples ?

File: ignite/metrics/maximum_mean_discrepancy.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["MaximumMeanDiscrepancy"]
+
+
+class MaximumMeanDiscrepancy(Metric):
+    r"""Calculates the mean of `maximum mean discrepancy (MMD)
+    <https://www.onurtunali.com/ml/2019/03/08/maximum-mean-discrepancy-in-machine-learning.html>`_.
```
Comment (vfdev-5) [inline ignite/metrics/maximum_mean_discrepancy.py:13]: Let's also put this link: https://jmlr.org/papers/volume13/gretton12a/gretton12a.pdf to the docstring

File: ignite/metrics/maximum_mean_discrepancy.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        v = self.var
+        XX = torch.exp(-0.5 * dxx / v)
+        YY = torch.exp(-0.5 * dyy / v)
+        XY = torch.exp(-0.5 * dxy / v)
+
+        # unbiased
+        n = x.shape[0]
+        XX = (XX.sum() - n) / (n * (n - 1))
+        YY = (YY.sum() - n) / (n * (n - 1))
+        XY = XY.sum() / (n * n)
```
Comment (vfdev-5) [inline ignite/metrics/maximum_mean_discrepancy.py:119]: If we were to implement mmd as kernel average over the full set, we can accumulate xx_sum, yy_sum, xy_sum and num_samples on `update` (e.g. `xx_sum += xx.sum()`) and in `compute`, run mmd computation:
```
mmd2 = (xx_sum - 2.0 * xy_sum + yy_sum).clamp(min=0.0) / num_samples
```
What do you think ?

Comment (kzkadc) [reply ignite/metrics/maximum_mean_discrepancy.py:125]: If we were to compute MMD over all samples, we would had to consider all $N^2$ combinations of k(x, y) from different batches.
I think it would be better to XX, YY, and XY within each batch and accumulate them.

File: ignite/metrics/maximum_mean_discrepancy.py (hunk 3 above)
Comment (kzkadc) [reply ignite/metrics/maximum_mean_discrepancy.py:24]: Thank you for the suggestion.

Although it is ideal to compute MMD using all $N$ samples, doing so requires to compute $N^2$ combinations of k(x, y), where we have to pool all inputs.
So, this metric computes MMD for each batch and takes the average.

I think using $B$ would be better so that users can be aware of this point.

File: ignite/metrics/maximum_mean_discrepancy.py (hunk 5 above)
Comment (vfdev-5) [reply ignite/metrics/maximum_mean_discrepancy.py:125]: OK, I think I get now your point about $N^2$ combinations. So, this metrics is computing average MMD over batch. Larger the batch size, closer "average MMD over batch" would be to MMD on N samples.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks a lot for the PR @kzkadc !
